test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java

```diff
@@ -0,0 +1,98 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/*
+ * @test
+ * @summary Tests that recursive locking doesn't cause excessive native memory usage
+ * @library /test/lib
+ * @run main TestRecursiveMonitorChurn
+ */
+public class TestRecursiveMonitorChurn {
+    static class Monitor {
+        volatile int i, j;
+        synchronized void doSomething() {
+            i++;
+            doSomethingElse();
+        }
+        synchronized void doSomethingElse() {
+            j++;
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
+        if (args.length == 1 && args[0].equals("test")) {
+            // The actual test, in a forked JVM.
+            for (int i = 0; i < 100000; i++) {
+                Monitor m = new Monitor();
+                m.doSomething();
+            }
+        } else {
+            ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(
+                    "-XX:+UnlockDiagnosticVMOptions",
+                    "-Xmx100M", "-XX:AsyncDeflationInterval=0",
+                    "-XX:NativeMemoryTracking=summary", "-XX:+PrintNMTStatistics",
+                    "TestRecursiveMonitorChurn",
+                    "test");
+            OutputAnalyzer output = new OutputAnalyzer(pb.start());
+            output.reportDiagnosticSummary();
+
+            output.shouldHaveExitValue(0);
+
+            // We want to see, in the final NMT printout, a committed object monitor size that is reasonably low.
+            // Like this:
+            // -           Object Monitors (reserved=208, committed=208)
+            //                             (malloc=208 #1) (at peak)
+            //
+            // Without recursive locking support, this would look more like this:
+            // -           Object Monitors (reserved=20800624, committed=20800624)
+            //                             (malloc=20800624 #100003) (at peak)
+
+            Pattern pat = Pattern.compile("- *Object Monitors.*reserved=(\\d+), committed=(\\d+).*");
+            boolean foundLine = false;
+            for (String line : output.asLines()) {
+                Matcher m = pat.matcher(line);
+                if (m.matches()) {
+                    long reserved = Long.parseLong(m.group(1));
+                    long committed = Long.parseLong(m.group(2));
+                    System.out.println(">>>>> " + line + ": " + reserved + " - " + committed);
+                    if (committed > 1000) {
+                        throw new RuntimeException("Allocated too many monitors");
+                    }
+                    foundLine = true;
```

Comment:
Small nit.
There is should be only one line like ' Object Monitors (reserved=20800624, committed=20800624)'
Some might just return true here? So no foundLine is needed.

Reply:
True. Fixed.